experiment/generate_tests.py

```diff
@@ -0,0 +1,233 @@
+                  test_suite, job):
+    """Returns a list of args fetched from the given fields."""
+
+    args = []
+    args.extend(get_args(job_name, common))
+    args.extend(get_args(job_name, cloud_provider))
+    args.extend(get_args(job_name, image))
+    args.extend(get_args(job_name, k8s_version))
+    args.extend(get_args(job_name, test_suite))
+    args.extend(get_args(job_name, job))
+    return args
+
+
+def for_each_job(job_name, common, cloud_providers, images, k8s_versions,
+                 test_suites, jobs):
+    '''Returns the list of envs and args for each the job with the given
+    job_name.'''
+
+    fields = job_name.split('-')
+    if len(fields) != 7:
+        raise ValueError('Expected 7 fields in job name', job_name)
+
+    cloud_provider_name = fields[3]
+    image_name = fields[4]
+    k8s_version_name = fields[5][3:]
+    test_suite_name = fields[6]
+
+    envs = generate_envs(job_name,
+                         common,
+                         cloud_providers[cloud_provider_name],
+                         images[image_name],
+                         k8s_versions[k8s_version_name],
+                         test_suites[test_suite_name],
+                         jobs[job_name])
+    args = generate_args(job_name,
+                         common,
+                         cloud_providers[cloud_provider_name],
+                         images[image_name],
+                         k8s_versions[k8s_version_name],
+                         test_suites[test_suite_name],
+                         jobs[job_name])
+    return envs, args
+
+
+def remove_generated_jobs(json_config):
+    for job_name, job_def in json_config.items():
+        tags = job_def.get('tags')
+        if tags is not None and 'generated' in tags:
+            del json_config[job_name]
+
+
+def main(json_config_path, yaml_config_path, output_dir):
+    '''Creates test job definitions for the jobs and their configs from the
+    yaml_config_path and writes them to json_config_path and output_dir.'''
+
+    # TODO(yguo0905): Validate the configurations from yaml_config_path.
+
+    with open(json_config_path) as fp:
+        json_config = json.load(fp)
+    remove_generated_jobs(json_config)
+
+    with open(yaml_config_path) as fp:
+        yaml_config = yaml.safe_load(fp)
+
+    # TODO(yguo0905): Create output_dir if it does not exist, or remove
+    # existing configs in output_dir otherwise.
+
+    for job_name, _ in yaml_config['jobs'].items():
+        # Get the envs and args for each job defined under "jobs".
+        envs, args = for_each_job(job_name,
+                                  yaml_config['common'],
+                                  yaml_config['cloudProviders'],
+                                  yaml_config['images'],
+                                  yaml_config['k8sVersions'],
+                                  yaml_config['testSuites'],
+                                  yaml_config['jobs'])
```

Comment:
Maybe this can be a list  under `yaml_config['config']` with arbitrary field name, and each job can point to any field it want to include, rather than solely depend on job names?